tests/backends/oracle/tests.py

```diff
@@ -61,3 +65,30 @@ def test_boolean_constraints(self):
             with self.subTest(field=field):
                 field.set_attributes_from_name('is_nice')
                 self.assertIn('"IS_NICE" IN (0,1)', field.db_check(connection))
+
+
+@unittest.skipUnless(connection.vendor == 'oracle', 'Oracle tests')
+class HiddenNoDataFoundException(TransactionTestCase):
+    available_apps = ['backends']
+
+    def test_hidden_no_data_found_exception(self):
+        # "ORA-1403: no data found" exception is hidden by Oracle OCI library
+        # when INSERT statement is used with RETURNING clause (see #28859).
+        with connection.cursor() as cursor:
+            # Create trigger that raise "ORA-1403: no data found".
+            cursor.execute("""
+                CREATE OR REPLACE TRIGGER "TRG_NO_DATA_FOUND"
+                AFTER INSERT ON "BACKENDS_SQUARE"
+                FOR EACH ROW
+                BEGIN
+                    RAISE NO_DATA_FOUND;
+                END;
+            """)
+        with self.assertRaisesMessage(DatabaseError, (
+            'The database did not return a new row id. Probably "ORA-1403: no '
+            'data found" has been raised internally, it is hidden by Oracle '
+            'OCI library (see #28859).'
+        )):
+            Square.objects.create(root=2, square=4)
+        with connection.cursor() as cursor:
```

Comment:
Should this be in a try/finally in case the assertion fails? I see some other test errors in that case.